File: tests/main/scram.test.ts
import { describe, it, expect } from 'vitest'
import { proofFor, serverSignatureFor, scramDigest } from '../../src/main/irc/scram'

/**
 * SCRAM, against the vector the RFC prints.
 *
 * The exchange is four HMACs and an XOR in a particular order, and getting any
 * of them slightly wrong fails in exactly one way: the server says 904 and the
 * user is told their password is wrong. This checked the primitives it did not
 * use before — `crypto.createHmac` returns 32 bytes, which was never in doubt
 * — so the code that actually runs was untested. RFC 7677 §3 knows the answer.
 */
describe('SCRAM-SHA-256 against RFC 7677', () => {
  const chosen = { password: 'pencil', digest: 'sha256' as const, keyLength: 32 }
  const salt = Buffer.from('W22ZaJ0SNY7soEsUEjb6gQ==', 'base64')
  const iterations = 4096

  const clientFirstBare = 'n=user,r=rOprNGfwEbeRWgbNEkqO'
  const serverFirst =
    'r=rOprNGfwEbeRWgbNEkqO%hvYDpWUa2RaTCAfuxFIlj)hNlF$k0,s=W22ZaJ0SNY7soEsUEjb6gQ==,i=4096'
  const clientFinalNoProof =
    'c=biws,r=rOprNGfwEbeRWgbNEkqO%hvYDpWUa2RaTCAfuxFIlj)hNlF$k0'
  const authMessage = `${clientFirstBare},${serverFirst},${clientFinalNoProof}`

  it('computes the client proof the RFC prints', () => {
    expect(proofFor(chosen, salt, iterations, authMessage).toString('base64')).toBe(
      'dHzbZapWIk4jUhN+Ute9ytag9zjfMHgsqmmiz7AndVQ='
    )
  })

  it('computes the server signature the RFC prints', () => {
    expect(
      serverSignatureFor(chosen, salt, iterations, authMessage).toString('base64')
    ).toBe('6rriTRBi23WpRR/wtup+mMhUZUn/dB5nLTJRsjl95G4=')
  })

  it('the channel binding in client-final is the "n,," we sent', () => {
    expect(Buffer.from('n,,').toString('base64')).toBe('biws')
  })
})

describe('SCRAM-SHA-512', () => {
  it('is the same exchange with a longer key', () => {
    expect(scramDigest('SCRAM-SHA-512')).toEqual({ digest: 'sha512', keyLength: 64 })
    expect(scramDigest('SCRAM-SHA-256')).toEqual({ digest: 'sha256', keyLength: 32 })
    expect(scramDigest('PLAIN')).toBeNull()
  })

  it('produces a proof of the right length, and a different one', () => {
    const salt = Buffer.from('W22ZaJ0SNY7soEsUEjb6gQ==', 'base64')
    const auth = 'n=user,r=abc,r=abcdef,s=x,i=4096,c=biws,r=abcdef'

    const sha256 = proofFor({ password: 'pencil', digest: 'sha256', keyLength: 32 }, salt, 4096, auth)
    const sha512 = proofFor({ password: 'pencil', digest: 'sha512', keyLength: 64 }, salt, 4096, auth)

    expect(sha256).toHaveLength(32)
    expect(sha512).toHaveLength(64)
  })
})

File: android/app/src/test/java/org/switchboard/android/ScramTest.kt
package org.switchboard.android

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.switchboard.android.irc.Scram

/**
 * SCRAM, against the vector the RFC prints.
 *
 * The exchange is four HMACs and an XOR in a particular order, and getting any
 * of them slightly wrong fails in exactly one way: the server says 904 and the
 * user is told their password is wrong. `tests/main/scram.test.ts` checks the
 * desktop against the same numbers, which is what makes two implementations of
 * the same handshake worth having.
 */
class ScramTest {

    private val serverFirst =
        "r=rOprNGfwEbeRWgbNEkqO%hvYDpWUa2RaTCAfuxFIlj)hNlF\$k0," +
            "s=W22ZaJ0SNY7soEsUEjb6gQ==,i=4096"

    @Test
    fun `computes the client proof RFC 7677 prints`() {
        val scram = Scram("user", "pencil", "SCRAM-SHA-256", nonce = "rOprNGfwEbeRWgbNEkqO")

        assertEquals("n,,n=user,r=rOprNGfwEbeRWgbNEkqO", scram.first())

        val clientFinal = scram.next(serverFirst)
        assertEquals(
            "c=biws,r=rOprNGfwEbeRWgbNEkqO%hvYDpWUa2RaTCAfuxFIlj)hNlF\$k0," +
                "p=dHzbZapWIk4jUhN+Ute9ytag9zjfMHgsqmmiz7AndVQ=",
            clientFinal
        )
    }

    @Test
    fun `accepts the server signature RFC 7677 prints, and refuses any other`() {
        val good = Scram("user", "pencil", "SCRAM-SHA-256", nonce = "rOprNGfwEbeRWgbNEkqO")
        good.first()
        good.next(serverFirst)
        assertNull(
            "a verified exchange has nothing left to send",
            good.next("v=6rriTRBi23WpRR/wtup+mMhUZUn/dB5nLTJRsjl95G4=")
        )
        assertTrue("and it says so", good.serverVerified)

        // SCRAM proves both sides knew the password. This is the half that was
        // computed and thrown away, so a server that could not prove it was
        // treated exactly like one that could.
        val bad = Scram("user", "pencil", "SCRAM-SHA-256", nonce = "rOprNGfwEbeRWgbNEkqO")
        bad.first()
        bad.next(serverFirst)
        assertNull(bad.next("v=AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="))
        assertFalse("a server that did not know the password is not verified", bad.serverVerified)
    }

    @Test
    fun `refuses a server that did not echo our nonce`() {
        val scram = Scram("user", "pencil", "SCRAM-SHA-256", nonce = "rOprNGfwEbeRWgbNEkqO")
        scram.first()
        assertNull(
            "an exchange we did not start is not one to answer",
            scram.next("r=somebodyelsesnonce,s=W22ZaJ0SNY7soEsUEjb6gQ==,i=4096")
        )
        assertFalse(scram.serverVerified)
    }

    @Test
    fun `SHA-512 is the same exchange with a longer proof`() {
        val scram = Scram("user", "pencil", "SCRAM-SHA-512", nonce = "rOprNGfwEbeRWgbNEkqO")
        scram.first()
        val clientFinal = scram.next(serverFirst)
        assertNotNull(clientFinal)

        val proof = clientFinal!!.substringAfter("p=")
        // 64 bytes of proof, base64
        assertEquals(88, proof.length)
    }
}
